lib/device_detector/os.rb

```diff
@@ -1,16 +1,152 @@
+require 'set'
+
 class DeviceDetector
   class OS < Parser
 
+    def name
+      os_info[:name]
+    end
+
+    def short_name
+      os_info[:short]
+    end
+
+    def family
+      os_info[:family]
+    end
+
+    def desktop?
+      DESKTOP_OSS.include?(family)
+    end
+
     def full_version
-      raw_version = super
+      raw_version = super.to_s.split('_').join('.')
+      raw_version == "" ? nil : raw_version
+    end
+
+    private
 
-      # This solution won't scale, but for now it will do
-      if raw_version && name == 'Mac'
-        raw_version.split('_').join('.')
+    def os_info
+      from_cache(['os_info', self.class.name, user_agent]) do
+        os_name = NameExtractor.new(user_agent, regex_meta).call
+        if os_name && short = DOWNCASED_OPERATING_SYSTEMS[os_name.downcase]
```

Comment:
Please surround the `short = ...` assignment with parentheses, in order to keep the intention clear.